경북대 BE_이도훈_1주차 과제(1, 2, 3단계)

2iedo opened the pull request:
### 코드 구현 시 고민했던 점

- HashMap의 경우 <Key, Data>값이 이미 있을 경우 같은 Key값에 입력하면 자동으로 늦게 입력한 값으로 변경되는데, 이는 업데이트 기능과 동일한데 어떻게 해야 할지 궁금합니다.(제출한 파일에는 HashMap이 아닐 경우도 가정하여 Update는 따로 구현하여 제출하였습니다.)
- 업데이트 기능의 어노테이션 중 @PostMapping과 @PutMapping 두 가지가 있었는데 , 이 두가지의 차이점을 잘 모르겠습니다.
- 데이터를 담는 DTO인 Product 클래스는 커밋 메시지 작성 시 어떤 태그로 작성해야할 지 고민입니다. 이것도 기능 추가(feat)로 봐야 할까요??  
- 예시 이미지에는 추가, 삭제, 선택 등의 버튼이 나타나 있는데, 이것들 또한 구현해야 하는지 궁금합니다!(일단은 목록 보여주기만 구현했습니다.)

### 그 외의 궁금한 점
- 구현한 기능이 제대로 작동하는 지 확인하기 위해 cmd의 curl 커맨드를 사용했는데, 이것 이외에 응답 방식을 확인할 수 있는 프로그램이 있나요??

2iedo committed:
feat : Product 레코드 추가
Long id, String name, int price, String imageUrl을 포함하며, record 형태로 작성되었다.

2iedo committed:
feat : 삭제 기능 추가
삭제를 담당하는 deleteProduct는 @DeleteMapping 어노테이션을 사용하였다.
key값인 id를 입력받으며, HashMap에 있는 key값에 해당하는 데이터를 제거한다.

2iedo committed:
feat : 업데이트 기능 추가
업데이트를 담당하는 updateProduct는 @PutMapping 어노테이션을 사용하였다.
key값인 id와 Product를 입력받아 업데이트를 실행한다.

2iedo committed:
feat : 조회 기능 추가
전체 데이터 조회 기능을 담당하는 viewAllProducts는 @GetMapping 어노테이션을 사용하였다.
HashMap에 있는 모든 데이터를 Collection 타입으로 반환한다.
key값이 id인 데이터 조회 기능을 담당하는 viewProduct는 @GetMapping 어노테이션을 사용한다.
key값에 해당하는 id를 입력받아 HashMap에 key값이 id인 데이터를 반환한다.

2iedo committed:
docs : README.md 업데이트
구현 사항 추가

2iedo committed:
docs : 2단계 구현사항 추가

2iedo committed:
feat : style.css 추가

2iedo committed:
feat : products.html 추가

2iedo committed:
feat : 불러오기 기능 변경
thymeleaf의 model에 데이터를 담아 테이블의 형태로 보여지도록 변경

2iedo committed:
feat : ProductDAO 클래스 구현
데이터베이스의 생성, 추가, 삭제, 수정, 조회 기능이 포함된다.

2iedo committed:
docs : step-3 구현사항 추가

2iedo committed:
feat : getConnection() 메소드 추가
url, user, password의 기본값을 정한다

2iedo committed:
feat : 해시맵에서 데이터베이스로 변경

2iedo committed:
feat : datasource 추가

2iedo committed:
feat : 데이터 추가 페이지 추가

2iedo committed:
feat : AdminController 클래스 추가
html파일에서 delete와 put이 작동하지 않아 post와 get만 사용하여 수정 및 제거 기능 구현

2iedo committed:
feat : property 추가
html 모드 및 cache를 허용하지 않게 설정

2iedo committed:
feat : menu.css 추가
add.html과 edit.html의 css파일

2iedo committed:
feat : 버튼 및 테이블 행 추가
생성 버튼, 수정 버튼, 삭제 버튼 및 actions 요소 추가

2iedo committed:
feat : edit.html 추가
데이터 수정 시 나타나는 화면으로, id의 중복을 방지하기 위해 id는 변경하지 못하게 함

2iedo committed:
feat : 버튼 및 요소 추가
생성, 삭제, 수정 버튼 추가

2iedo committed:
feat : 조회 기능 수정

khj68 commented:
> ### 코드 구현 시 고민했던 점
> * HashMap의 경우 <Key, Data>값이 이미 있을 경우 같은 Key값에 입력하면 자동으로 늦게 입력한 값으로 변경되는데, 이는 업데이트 기능과 동일한데 어떻게 해야 할지 궁금합니다.(제출한 파일에는 HashMap이 아닐 경우도 가정하여 Update는 따로 구현하여 제출하였습니다.)
의도하신 비즈니스 로직이, 이미 존재할 때 저장이 안되도록 하려면 해당 상황에 에러처리를 해주시면 되겠습니다.
> * 업데이트 기능의 어노테이션 중 @PostMapping과 @PutMapping 두 가지가 있었는데 , 이 두가지의 차이점을 잘 모르겠습니다.
post는 새로 등록하는 경우, put은 이미 존재하는 것을 수정하는 경우입니다.
> * 데이터를 담는 DTO인 Product 클래스는 커밋 메시지 작성 시 어떤 태그로 작성해야할 지 고민입니다. 이것도 기능 추가(feat)로 봐야 할까요??
feat으로 봐주시면 되겠습니다. dto도 중요한 기능 중 하나입니다.
> * 예시 이미지에는 추가, 삭제, 선택 등의 버튼이 나타나 있는데, 이것들 또한 구현해야 하는지 궁금합니다!(일단은 목록 보여주기만 구현했습니다.)
이는 프로젝트 진행 관련으로 주최 측에 질문주시면 감사하겠습니다 :)
> 
> ### 그 외의 궁금한 점
> * 구현한 기능이 제대로 작동하는 지 확인하기 위해 cmd의 curl 커맨드를 사용했는데, 이것 이외에 응답 방식을 확인할 수 있는 프로그램이 있나요??
저는 실무에서 postman을 편리하게 사용하고 있습니다.

khj68 commented on src/main/java/gift/Application.java:
```diff
@@ -3,9 +3,19 @@
         SpringApplication.run(Application.class, args);
     }
+
+    public static Connection getConnection() throws Exception {
```
이 함수는 어디서 사용하고 있나요?

2iedo replied:
아 저번 클론코딩 때 한 코드 참고하여 작성하다 잘못 넣은 것 같습니다. 제거했습니다!

khj68 commented on src/main/java/gift/Controller/ProductController.java:
```diff
@@ -0,0 +1,68 @@
+            productDAO.deleteProduct(id);
+            return ResponseEntity.status(HttpStatus.OK).body("제거 성공");
+        } catch (Exception e) {
+            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("제거 실패");
```
모든 컨트롤러 메소드에서 try/catch 를 하는 것보다 global exception handler 도입을 추천드립니다.

2iedo replied:
넵 알겠습니다!

khj68 reviewed (changes requested):
전반적으로 클래스 분리도 잘되어있고 깔끔한 코드입니다.
다만 try/catch 와 같이 보일러플레이트처럼 반복되는 부분은 global하게 관리하는 방법을 고려해주세요.
고생하셨습니다.

ajy9851 committed:
전남대 BE_안재영_1주차 과제(1단계) (#11)

* docs(README.md): 구현 기능 목록 작성

* feat(Product.java): 상품 클래스 작성

요구사항 구현을 위한 상품 클래스

* feat(ProductDatabase.java): 데이터베이스 역할을 하는 클래스 생성

데이터베이스 없이 구현하는 것이 요구사항이니까 메모리에 저장하는 역할을 하는 클래스 생성

* fix(Product): 변수 명 수정

변수의 시작이 대문자로 되어 있어서 수정

* feat(ProductController): 컨트롤러 추가

RESTful 웹 서비스의 컨트롤러

ajy9851 committed:
전남대 BE_안재영_1주차 과제(2단계) (#149)

* docs(README.md): 구현 기능 목록 작성

* feat(Product.java): 상품 클래스 작성

요구사항 구현을 위한 상품 클래스

* feat(ProductDatabase.java): 데이터베이스 역할을 하는 클래스 생성

데이터베이스 없이 구현하는 것이 요구사항이니까 메모리에 저장하는 역할을 하는 클래스 생성

* fix(Product): 변수 명 수정

변수의 시작이 대문자로 되어 있어서 수정

* feat(ProductController): 컨트롤러 추가

RESTful 웹 서비스의 컨트롤러

* feat(ThymeleafProductController.java): 타임리프를 이용하기 위한 컨트롤러 작성

2단계 미션을 수행하기 위해 1단계와 다른 컨트롤러를 만들었습니다.

* feat(application.yml): HTML폼에서 PUT, DELETE를 사용하기 위해 작성

어떻게 HTML 폼 전송 시 PUT, DELETE 요청을 할지 몰라서 구글링 중 해당 파일로 설정을 해야한다해서 적용했습니다.

* feat(create-product.html): 제품 제작 html 작성

* feat(product-list): 제품 리스트 html 작성

* feat(product-list): 제품 수정  html 작성

* docs(README.md): 2단계 미션 설명 추가

---------

Co-authored-by: Morgan <51692363+gunom@users.noreply.github.com>

gunom committed:
Revert "전남대 BE_안재영_1주차 과제(2단계) (#149)" (#213)

This reverts commit 768aff6b763f4250a4ad2b8a7ea2cec38ae681f3.

gunom committed:
Revert "전남대 BE_안재영_1주차 과제(1단계) (#11)" (#214)

This reverts commit 6a08787306218d55296f1c908aa649fe173da630.

Youngini committed:
경북대 BE_김영인_1주차 과제(1단계) (#64)

* docs(Readme.md) : 기능 요구

* docs(README.md) : 기능 요구사항 분석 및 기능 목록 양식 수정

양식 수정

* [feat] product entity 구성

* [docs] product entity 구성

* [feat] product DB 구성

* [docs] product 전체 조회 기능 추가

* [feat] product 전체 조회 기능 구현

* [build] product 추가 기능 프로그래밍 요구 사항에 맞춰 수정

1. ProductController의 createProduct 메서드에서 validateProductDto 메서드를 분리하여 indent depth를 2 이하로 유지
2. 3항 연산자 대신 if-return 패턴을 사용
3. createProduct 메서드의 길이가 15라인 이하가 되도록 함
4. createProduct 메서드는 상품 생성 기능만 수행하도록 했습니다.
5. else 예약어를 사용하지 않고 if 문에서 값을 바로 반환하는 방식으로 구현

* [feat] product 전체 조회 기능 구현

* [feat] 특정 상품 조회하는 api 구현

* [docs] 특정 상품 조회하는 api 구현

* [feat] 특정 상품 업데이트하는 api 구현

* [feat] 특정 상품 삭제하는 api 구현

* [build] 전체 상품 조회 api 예외 처리 수정

* [build] 상품 정보 updtae와 상품 삭제의 id 조회 실패 시 에러 메세지 수정

* [docs] 상품 api 오류 메시지 수정

* [build] 상품 추가 api 오류 처리 수정

* [build] 상품에 image 추가하면서 관련 코드 모드 수정

* [docs] 2단계 과제를 위한 구현할 기능 목록 정리

* [feat] 상품 리스트를 보여주는 메인 페이지 완성

부트스트랩을 사용하여 꾸밈

* [feat] 상품 상세 설명 페이지 구성

* [feat] 상품 상세 설명 페이지 삭제 버튼 모달 창 완성

* [build] 상품 목록 페이지, 상품 상세 정보 페이지 용어 및 js 동작 수정

* [build] 상품 상세 정보 페이지 상단 상품 목록 버튼 클릭시, main으로 이어지도록 링크 수정

* [feat] 상품 업데이트 페이지 완성

* [docs] 상품 업데이트 페이지 완성

* build(main.html) : 상품 목록 페이지 주석 추가

* feat(productAdd.html) : 상품 추가 페이지 완성

* feat(productAdd.html) : 상품 추가 페이지 완성

* build(productAdd.html) : 상품 추가 페이지 꾸미지

* docs(README.md) : 3단계 과제 구현할 기능 목록 정리

* feat(application.properties) : H2 데이터베이스 사용하기 위한 properties 설정

* feat(application.properties) : H2 데이터베이스 사용하기 위한 properties 설정 수정

spring.sql.init.mode=always:
애플리케이션 시작 시 항상 데이터베이스 초기화 스크립트를 실행합니다.

spring.sql.init.schema-locations=classpath:schema.sql:
데이터베이스 스키마 초기화 스크립트의 위치를 지정합니다.

classpath:schema.sql 파일에 정의된 SQL 스크립트가 실행됩니다.

spring.sql.init.data-locations=classpath:data.sql:
데이터베이스 초기 데이터 삽입 스크립트의 위치를 지정합니다.

classpath:data.sql 파일에 정의된 SQL 스크립트가 실행됩니다.

* feat(.sql) : H2 데이터베이스 테이블 구조 정의 및 생성

* feat(Repository) : DB의 CRUD를 구현하기 위한, 인터페이스와 repository 구현

* feat(ProductService) : productservice의 create 코드 H2 DB 사용하도록 수정

* feat(ProductService) : productservice의 update 코드 H2 DB 사용하도록 수정

* feat(ProductService) : productservice의 getProduct 코드 H2 DB 사용하도록 수정

* feat(ProductService) : productservice의 updateProduct 코드 H2 DB 사용하도록 수정

* feat(ProductService) : productservice의 deleteProduct 코드 H2 DB 사용하도록 수정

* build(ProductService) : 더이상 사용하지 않는 모듈의 import 문 삭제

* docs(README.md) : 구현 및 수정 완료한 기능 목록 표시

* build(productService) : service 메소드 내 예외처리 코드 분리

* build(productService) : service 메소드 내 예외처리 코드 다른 폴더로 분리

* docs(README/md) : postman과 swagger 사용한 api 작동 확인

khj68 commented:
도훈님 conflict resolve 부탁드립니다!

2iedo committed:
Merge branch 'main' into 2iedo

2iedo commented:
> 도훈님 conflict resolve 부탁드립니다!

넵 해결했습니다!
